### ISSUE TITLE
One class is not detected for custom dataset

### ISSUE THREAD

**Issue by nlm-yuh5** [ultralytics/yolov5#2870]
## ❔Question

I trained my custom dataset for 2-class detection problem (uninfected cell VS infected cells). However, the trained model can only detect one of the class(uninfected cells). 

Granted in my dataset there is an data imbalance problem. For the infected cell class, there is only 1,142 instances in the dataset, while there are 33,071 instances for the uninfected class. But I don't think that it would cause the trained model to detect nothing for the small number class (infected cells).

Do you have an idea of why this would happen?

## Additional context

![device-2021-04-08-224137](https://user-images.githubusercontent.com/63370208/115484694-39aa1380-a221-11eb-88d4-bdcf3735c287.png)

This is an example of detection. I programmed the app to draw yellow bbox for uninfected cells, and red for infected cells. This image do have infected cells, but as you can see they are all detected as uninfected.

**Comment by glenn-jocher** [comment]
@nlm-yuh5 👋 Hello! Nice app! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/

**Comment by glenn-jocher** [comment]
@nlm-yuh5 what development pathway did you use for your app by the way, it looks really good!

@SergioSanchezUltralytics see app example above.

**Comment by nlm-yuh5** [comment]
@glenn-jocher Thanks for the speedy reply! I will try to provide more details.

We have well-annotated dataset for this project. I trained the model with the default setting except that I used a larger image size. As far as the Android app, it's just a slight modification from the PyTorch [Android demo app](https://github.com/pytorch/android-demo-app/tree/master/ObjectDetection).

**Comment by nlm-yuh5** [comment]
@glenn-jocher I'm putting a [link ](https://drive.google.com/drive/folders/1vWLQLVyvfTKy2oXP7gFr6LY_4uwZpJ6p?usp=sharing)here for the results, hopefully they give the details.

**Comment by glenn-jocher** [comment]
@nlm-yuh5 I see! We generally don't provide personal feedback on custom dataset training, but it's rather clear from your training results that your app will not work well. You should spend some time analyzing the folder and following the above training best practices recommendations.

I see you are a commercial/government user, if you are interested in a paid support plan and guidance on improving your results you can reach out to us at https://ultralytics.com/contact.

**Comment by Yu-Hang** [comment]
@glenn-jocher is there an easy way to substitute the loss function?

**Comment by glenn-jocher** [comment]
@Yu-Hang yes the loss function can be modified here in utils/loss.py:
https://github.com/ultralytics/yolov5/blob/5f7d39fede4de8af98472bd009c63c3a86568e2d/utils/loss.py#L88-L162
